addons/mail/models/discuss/ir_websocket.py

```diff
@@ -21,6 +21,47 @@ def _get_im_status(self, data):
             )]
         return im_status
 
+    def _get_missed_presences_identity_domains(self, presence_channels):
+        identity_domain = super()._get_missed_presences_identity_domains(presence_channels)
+        if guest_ids := [
+            g.id for g, _ in presence_channels if isinstance(g, self.pool["mail.guest"])
+        ]:
+            identity_domain.append([("guest_id", "in", guest_ids)])
+        return identity_domain
+
+    @add_guest_to_context
+    def _build_presence_channel_list(self, presences):
+        channels = super()._build_presence_channel_list(presences)
+        # sudo - mail.guest: guest presences can be sent to anyone.
+        guests = (
+            self.env["mail.guest"]
+            .sudo()
+            .search([("id", "in", [int(p[1]) for p in presences if p[0] == "mail.guest"])])
+        )
+        channels.extend((guest, "presence") for guest in guests)
+        guest = self.env["mail.guest"]._get_guest_from_context()
+        if (
+            guest
+            and (not self.env.user or self.env.user._is_public())
+            or self.env.user
+            and self.env.user._is_portal()
+        ):
+            # sudo - discuss.channel: portal or guests can access their own channels.
+            channel_ids = (
+                guest.sudo().channel_ids.ids if guest else self.env.user.partner_id.channel_ids.ids
+            )
```

Comment:
I think this would incorrectly prioritize guest if the current user is portal with a guest cookie.

We're better off separating the conditions more clearly, especially since the `if` is harder to read than it should.